adoc/extensions/sycl_khr_work_item_queries.adoc

```diff
@@ -0,0 +1,169 @@
+[[sec:khr-work-item-queries]]
+= SYCL_KHR_WORK_ITEM_QUERIES
+
+This extension allows developers to access instances of the
+[code]#sycl::nd_item#, [code]#sycl::group# and [code]#sycl::sub_group# classes
+without having to explicitly pass them as arguments to each function used on the
+device.
+
+{note} Passing such instances as arguments can result in a clearer interface
+that is less error-prone to use.
+For example, if a function accepts a [code]#sycl::group#, the caller must assume
+that function may call a [code]#sycl::group_barrier# and ensure that associated
+control flow requirements are satisfied.
+It is recommended that this extension is used only when modifying existing
+interfaces is not feasible.
+{endnote}
+
+[[sec:khr-work-item-queries-dependencies]]
+== Dependencies
+
+This extension has no dependencies on other extensions.
+
+[[sec:khr-work-item-queries-feature-test]]
+== Feature test macro
+
+An implementation supporting this extension must predefine the macro
+[code]#SYCL_KHR_WORK_ITEM_QUERIES# to one of the values defined in the table
+below.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+[[sec:khr-work-item-queries-free-funcs]]
+== New free functions to access instances of special SYCL classes
+
+This extension adds the following free functions to the
+[code]#sycl::khr::this_work_item# namespace, which provide information about the
+currently executing work-item.
+
+It is the user's responsibility to ensure that these functions are called in a
+manner that is compatible with the kernel's launch parameters, as detailed in
+the definition of each function.
+Calling these functions from an incompatible kernel results in undefined
+behavior.
+
+'''
+
+.[apidef]#khr::this_work_item::get_nd_item#
+[source,role=synopsis,id=api:khr-this-work-item-get-nd-item]
+----
+namespace sycl::khr::this_work_item {
+
+template <int Dimensions>
+nd_item<Dimensions> get_nd_item();
+
+}
+----
+
+_Preconditions_: [code]#Dimensions# must match the dimensionality of the
+currently executing kernel.
+The currently executing kernel must have been launched with a
+[code]#sycl::nd_range# argument.
+
+_Returns_: A [code]#sycl::nd_item# instance representing the calling work-item
+in the [code]#sycl::nd_range#.
+
+'''
+
+.[apidef]#khr::this_work_item::get_work_group#
+[source,role=synopsis,id=api:khr-this-work-item-get-work-group]
+----
+namespace sycl::khr::this_work_item {
+
+template <int Dimensions>
+group<Dimensions> get_work_group();
```

Comment:
I didn't want to introduce a dependency, because people might want to use these functions without using the new group types.

#638 defines an implicit conversion from `sycl::group` to `khr::work_group`, so the way things are written here permits both of the following options:
```c++
sycl::group g = sycl::khr::this_work_item::get_work_group();
sycl::work_group wg = sycl::khr::this_work_item::get_work_group();
```

The conversion from `khr::work_group` to `sycl::work_group` is explicit, so it wouldn't be as nice to use if we flipped things around.

Defining just one function seemed preferable to me compared to defining both `get_group()` and `get_work_group()`.  `get_work_group()` is the name of the function that we would be adding to SYCL-Next if the KHR is successful.

Reply:
> `get_work_group()` is the name of the function that we would be adding to SYCL-Next if the KHR is successful.

And what type will this function return?  What if #638 is also adopted into SYCL-Next?  What will be the return type of `get_work_group()`?

Reply:
My assumption was that we'd keep the function called `get_work_group()` even if we decide not to rename the `group` class via #638, but you're right that might not be a safe assumption. I've renamed this to `get_group()` in 31dc1c21ae73533979bebb520c343cd076b734af.

I didn't add a `get_work_group()` because I'm not sure how we would handle the dependency; it's not immediately clear to me which extension would be responsible for defining a `get_work_group()` function that returns `khr::work_group`, and I'm reluctant to put us in a situation where we cannot release either KHR until they're both ready.

Do you think it would make sense to hold off on introducing `get_work_group()` until both KHRs are merged?  We could then decide whether to introduce `get_work_group()` via a revision to one of the KHRs, or to wait until SYCL-Next if we're confident both KHRs will be promoted.

Reply:
Here's the progression I see:

Now: Both KHRs are independent:

* This KHR has `khr::this_work_item::get_group` returning `sycl::group`.  If an application wants to use both KHRs together, they can do:

  ```
  khr::work_group<1> g = khr::this_work_item::get_group<1>();  //  works because of conversion
  ```

If this KHR is adopted into SYCL-Next before #638:

* The API is named `sycl::this_work_item::get_group` and still returns `sycl::group`.  If an application wants to use this API together with #638, they can still use the code sequence above.

*  If #638 is later adopted into SYCL-Next, we also add `sycl::this_work_item::get_work_group` returning `sycl::work_group`.

If #638 is adopted into SYCL-Next before this KHR:

* Applications using this KHR can continue to use the sequence shown above, or they can do this:

  ```
  sycl::work_group<1> g = khr::this_work_item::get_group<1>();  //  works because of conversion
  ```

* If this KHR is later adopted into SYCL-Next, we also add `sycl::this_work_item::get_work_group` returning `sycl::work_group`.  We might decide *not* to adopt `sycl::this_item::get_group` in this case.